source/Engine.cpp

```diff
@@ -1461,10 +1473,23 @@ void Engine::CalculateStep()
 
 	// Keep track of the flagship to see if it jumps or enters a wormhole this turn.
 	const Ship *flagship = player.Flagship();
+	bool flagshipWasCloaked = (flagship && flagship->Cloaking() == 1);
 	bool wasHyperspacing = (flagship && flagship->IsEnteringHyperspace());
 	// Move all the ships.
 	for(const shared_ptr<Ship> &it : ships)
+	{
+		bool wasCloaked = it->Cloaking() == 1;
 		MoveShip(it);
+		// If we decloaked, and we're in the same system as the player, they Encounter us
+		if(wasCloaked && it->Cloaking() < 1 && flagship->GetSystem() == it->GetSystem()
+			&& flagship->Cloaking() < 1 && it.get() != flagship)
+				eventQueue.emplace_back(player.FlagshipPtr(), it, ShipEvent::ENCOUNTER);
+	}
+	// If we *are* the player, and we decloaked, we encounter everyone
+	if(flagshipWasCloaked && flagship->Cloaking() < 1)
+		for(const shared_ptr<Ship> &ship : ships)
+			if(ship->GetSystem() == playerSystem && ship->Cloaking() < 1)
+				eventQueue.emplace_back(player.FlagshipPtr(), ship, ShipEvent::ENCOUNTER);
```

Comment:
I'm not entirely sure I follow. There needs to be a whole separate loop for the flagship to encounter every ship in the system, wherever we put it, doesn't there?

Unless you're saying we pull `MoveShip()` for the flagship _out_ of the loop above, so that we know its status (and then skip it within the loop), and then if it uncloaked we can encounter each ship in the loop...?

Reply:
> Unless you're saying we pull MoveShip() for the flagship out of the loop above, so that we know its status (and then skip it within the loop), and then if it uncloaked we can encounter each ship in the loop...?

yup that's what I meant

Reply:
OK! Implemented; please do check it over to make sure I didn't screw up something in the refactor. (It's late here.)